rfcs/generic-functions.md

```diff
@@ -0,0 +1,110 @@
+# generic functions
+
+## Summary
+
+Extend the syntax and semantics of functions to support explicit generic functions, which can bind type parameters as well as data parameters.
+
+## Motivation
+
+Currently Luau allows generic functions to be inferred but not given explicit type annotations. For example
+
+```
+function id(x) return x end
+local x: string = id("hi")
+local y: number = id(37)
+```
+
+is fine, but there is no way for a user to write the type of `id`.
+
+## Design
+
+Allow functions to take type parameters as well as function parameters, similar to Java/Typescript/...
+
+```
+function id<a>(x : a) : a return x end
+```
+
+Functions may also take generic type pack arguments for varargs, for instance:
+
+```
+function compose<a...>(... : a...) -> (a...) return ... end
+```
+
+This change is *not* only syntax, as explicit type parameters need to be part of the semantics of types. For example:
+
+```
+function f()
+  function id(x) return x end
+  return id
+end
+function g()
+  local y
+  function oh(x)
+    if not(y) then y = x end
+    return y
+  end
+  return oh
+end
+```
+
+The types of these functions are
+
+```
+  f : () -> <a>(a) -> a
+  g : <a>() -> (a) -> a
+```
+
+so this is okay:
```

Comment:
I was playing around with this to make sure I understood, but it seems like this example gives an error, at least in Studio today. It infers `i` as `string -> string` and then trying to pass 37 causes an error.

I made some other examples, like a memo function, and a lazy/thunk function, and they ran into that same diagnostic. Is there something I'm missing here?

Reply:
Oh that's interesting you need to put a type annotation on `i` to get this to work, since by default Luau infers the monomorphic type for `i`, not the polymorphic type. I updated the example so it now works in Studio.